### DISCUSSION TITLE
building etcd - branch vs tag

### DISCUSSION THREAD

**Discussion by dulhaver** [etcd-io/etcd#17894] [Q&A] [answered]
I see these instruction in then [docs for building etcd from sources](https://etcd.io/docs/v3.5/install/)

`git clone -b v3.5.12 https://github.com/etcd-io/etcd.git`

despite that I see branch names like `release-3.5`. so `v3.5.12` seems to be a tag.
The clone command works but still leaves me with an error like
```
You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:
...
```

Is that how it is supposed to be? 
if so ... can anybody kindly explain why not cloning the branch `release-3.5` (which might be the 'stable' version 3.5 I am guessing) but the tag `v3.5.12` ?

they seem to differ quite a lot

**Comment by ivanvc** [comment]
Hi @dulhaver, the latest 3.5 release is `v3.5.13`. The website needs to be updated. It was recently released. There's active work happening on the `release-3.5` branch; if you pull the `HEAD,` you may have a pre-release version of the 3.5 branch.

Once the [release criteria is met](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/release.md#patch-release-criteria) for a supported version (3.4 and 3.5), it will be released (and the tag will point to that specific commit).

You should check out a specific tag, as you may get a version with unstable functionality (unlikely, as these two branches get only bug fixes, backports, and dependency updates). Still, if you're building from the sources, that's the only way to know you're running what was released in that specific version.

**Reply by dulhaver** [reply]
thx @ivanvc 

sorry, I do not wuite understand the essence of your advice

> _You **should check out a specific tag**, as you **may get a version with unstable** functionality_ 

should one check out the `release-x.y` branch vs check out a specific tag `x.y.z` for building for production? My understanding of the term `release` kind of suggest it is good for the public (aka stable). Is that a miss interpretation of a non-mothertounge speaker?

Maybe it's lacking understanding of git
> _if you pull the **HEAD**, you may have a pre-release version of the 3.5 branch._

`HEAD` would be `git clone -b release-3.5` ?

**Reply by dulhaver** [reply]
I think I start getting it.

- branch `release-3.5` is the development branch of version `3.5.`
- it contains all it's stable releases marked be tags (`v3.5.11`, `v3.5.12`, etc ...)
- the `HEAD` of each release branch is stable for the exact one commit tagged as `v3.5.x`. everything after thtat point in time is by definition development code, aka potentially unstable

so you would need to ... 
1. define the supported version (`3.4`, `3.5`) you want to use
2. identify that latest tag of that version
3. clone that exact tag & build with that code

is that a fair understanding?

**Reply by ivanvc** [reply, marked as answer]
> `HEAD` would be git clone -b release-3.5?

Yes, `HEAD` is the latest commit in a branch. After you clone, a `git pull origin release-3.5` would also bring the latest `HEAD`.

> is that a fair understanding?

Yes, you're correct. I also forgot to mention that the following message from Git:

> You are in 'detached HEAD' state. You can look around, make experimental
> changes and commit them, and you can discard any commits you make in this
> state without impacting any branches by switching back to a branch.
> ..,

It is not an error. It states precisely that you're on a specific commit (because tags point to commits).

**Comment by dulhaver** [comment]
great, thanks a ton @ivanvc
